git/git.go

```diff
@@ -40,6 +40,22 @@ type Ref struct {
 	Sha  string
 }
 
+func (r *Ref) NameOnRemote(remotename string) string {
+	if r.Type != RefTypeLocalBranch {
+		return r.Name
+	}
+
+	if len(remotename) == 0 {
+		return r.Name
+	}
+
+	if RemoteForBranch(r.Name) != remotename {
+		return r.Name
+	}
+
+	return RemoteBranchForLocalBranch(r.Name)
+}
+
```

Comment:
I think this is fine for auto selection which is what the function is designed to do, but yeah if we want to support full syntax this may be overridden; the problem is I'm not sure we get the remote-side name in the pre-push hook?

I'm somewhat skeptical of the actual usefulness of this ref in the api anyway except as a convenience hint for early rejection. It's so easy to game it's of no real use security wise.